lib/stdlib/src/erl_expand_records.erl

```diff
@@ -335,9 +373,33 @@ expr({record_index,Anno,Name,F}, St) ->
     expr(I, St);
 expr({record,Anno0,Name,Is}, St) ->
     Anno = mark_record(Anno0, St),
-    expr({tuple,Anno,[{atom,Anno0,Name} |
-                      record_inits(record_fields(Name, Anno0, St), Is)]},
-         St);
+    
+    IsUndefined = [{RF, AnnoRF, Field, {atom, AnnoRF, 'undefined'}} || {record_field=RF, AnnoRF, Field, _} <- Is],
+    Fields = lists:flatten(lists:sort([atom_to_list(FieldAtom) || {record_field, _, {atom, _, FieldAtom}, _} <- Is])),
+    R_default_init = [{atom,Anno,Name} |
+                       record_inits(record_fields(Name, Anno0, St),IsUndefined)],
+    R_init = [{atom,Anno,Name} |
+              record_inits(record_fields(Name, Anno0, St), Is)],
+    Vars = lists:flatten(traverse_af(Is, fun save_vars/2)),
+    %% If R_init contains free variables that was not bound via Is
+    case free_variables(R_init, Vars) of
+        true ->
+            FName = list_to_atom("erl_expand_records_init_"++atom_to_list(Name)++"_"++Fields),
```

Comment:
This might end up blowing the atom character limit, and could in theory clash with something that the user has provided.

How about using a mangled identifier with a sequence number instead, a la list comprehensions and funs?